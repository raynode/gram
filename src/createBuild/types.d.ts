import { GraphQLSchema, GraphQLType } from 'graphql'
import { IFieldResolver, IResolvers, ITypeDefinitions } from 'graphql-tools'

import { GQLBUILDER } from '../types/constants'

import { createAddType } from './method-addType'

// Context == GraphQLContext ({ user: {â€¦}, db: DBConnection })
// BuildMode == BuildMode of the build ('user' | 'admin')

export interface TypeDefs {
  typeDefs: ITypeDefinitions
  resolvers: IResolvers
}
export interface Build<BuildMode> {
  type: typeof GQLBUILDER
  buildMode: BuildMode
  addQuery: AddResolvable
  addMutation: AddResolvable
  addSubscription: AddResolvable
  addType: ReturnType<typeof createAddType>
  toSchema: () => GraphQLSchema
  toTypeDefs: () => TypeDefs
}

export type GQLRecord = Record<string, string>
export type Resolvers = IResolvers
export type BuildModeGenerator<BuildMode, Result> = (
  buildMode?: BuildMode,
) => Result
export type BuildModeArgsGenerator<BuildMode, Args, Result = void> = (
  buildModeGenerator: BuildModeGenerator<BuildMode, Args>,
) => Result
export type FieldType = string | GraphQLType
export type Fields = Record<string, FieldType>

export type Resolvables = 'Query' | 'Mutation' | 'Subscription'
export interface ObjectTypesRecordConfig {
  fields: GQLRecord
  interface: string
}
export interface InterfaceTypesRecordConfig {
  fields: GQLRecord
}
export interface InputTypesRecordConfig {
  fields: GQLRecord
}
export interface EnumTypesRecordConfig {
  values: string[]
}
export interface CreateableTypesRecord {
  scalar: string[]
  type: Record<string, ObjectTypesRecordConfig>
  interface: Record<string, InterfaceTypesRecordConfig>
  input: Record<string, InputTypesRecordConfig>
  enum: Record<string, EnumTypesRecordConfig>
}
export type CreateableTypes = keyof CreateableTypesRecord

export interface AddResolvableConfig<Source, Context, Resolver> {
  args?: GQLRecord
  resolver?: Resolver
}
export type AddResolvable = (<Source, Context>(
  name: string,
  type: FieldType,
  config?: any,
) => void) &
  (<BuildMode, Source, Context, FieldType>(
    name: string,
    type: BuildModeGenerator<BuildMode, FieldType>,
    config?: any,
  ) => void)

export type AddResolver<Context> = <Source>(
  base: string,
  name: string,
  config: AddResolvableConfig<Source, Context, IFieldResolver<Source, Context>>,
) => void
