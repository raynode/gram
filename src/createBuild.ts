import { GraphQLType } from 'graphql'
import {
  IFieldResolver,
  IResolvers,
  makeExecutableSchema,
} from 'graphql-tools'
import { isEmpty, reduce } from 'lodash'

type GQLRecord = Record<string, string>
type Resolver<Source, Context> = IFieldResolver<Source, Context>
type Resolvers = IResolvers
type BuildModeGenerator<BuildMode, Result> = (buildMode?: BuildMode) => Result

const isBuildModeGenerator = <BuildMode, Result>(
  val: any,
): val is BuildModeGenerator<BuildMode, Result> => typeof val === 'function'

export const typeToString = <Type>(type: Type) => {
  return type.toString()
}

const buildData = (data: GQLRecord) =>
  reduce(
    data,
    (list, type, name) => {
      list.push(`${name}: ${type}`)
      return list
    },
    [],
  ).join('\n')

const generateData = (typeName: string, data: GQLRecord) => `
  type ${typeName} {
    ${buildData(data)}
  }
`

const generateNonEmpty = (typeName: string, data: GQLRecord) =>
  isEmpty(data) ? '' : generateData(typeName, data)

export const generateTypeDefs = (queries, mutations, subscriptions, types) => {
  if (isEmpty(queries)) throw new Error('Query cannot be empty')

  const typeDefs = `
    ${generateData('Query', queries)}
    ${generateNonEmpty('Mutation', mutations)}
    ${generateNonEmpty('Subscription', subscriptions)}
  `
  return typeDefs
}

export const createBuild = <BuildMode = null, Context = any>(
  buildMode?: BuildMode,
) => {
  const queries: GQLRecord = {}
  const mutations: GQLRecord = {}
  const subscriptions: GQLRecord = {}
  const types: Record<string, GQLRecord> = {}
  const resolvers: Resolvers = {}
  const addResolver = <Source>(
    base: string,
    name: string,
    resolver: Resolver<Source, Context>,
  ) => {
    resolvers[base] = resolvers[base] || {}
    resolvers[base][name] = resolver
  }

  type AddQuery = (<Source, Type = string | GraphQLType>(
    name: string,
    type: Type,
    resolver?: Resolver<Source, Context>,
  ) => void) &
    (<Source, Type = string | GraphQLType>(
      name: string,
      type: BuildModeGenerator<BuildMode, Type>,
      resolver?: BuildModeGenerator<BuildMode, Resolver<Source, Context>>,
    ) => void)

  const addQuery: AddQuery = (name, type, resolver?) => {
    if (isBuildModeGenerator<BuildMode, string | GraphQLType>(type)) {
      const x = type(buildMode)
      const y = typeToString(x)
      queries[name] = y
      addResolver('Query', name, resolver(buildMode))
    } else {
      queries[name] = typeToString(type)
      addResolver('Query', name, resolver)
    }
  }
  const builder = {
    builder: 'Build',
    buildMode,
    addQuery,
    addMutation: () => null,
    addSubscription: () => null,
    addType: () => null,
    toSchema: () => makeExecutableSchema(builder.toTypeDefs()),
    toTypeDefs: () => {
      const typeDefs = generateTypeDefs(
        queries,
        mutations,
        subscriptions,
        types,
      )
      return { typeDefs, resolvers }
    },
  }
  return builder
}
