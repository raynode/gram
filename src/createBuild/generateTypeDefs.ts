import { ITypeDefinitions } from 'graphql-tools'
import { isEmpty, map, reduce } from 'lodash'
import {
  CreateableTypeConfig,
  CreateableTypes,
  CreateableTypesRecord,
  EnumTypesRecordConfig,
  GQLRecord,
  Resolvables,
  ResolvablesRecord,
} from './types'
import { reduceRecord } from './utils'

const generateCreateableType = <Source, Context>(
  createable: string,
  typeName: string,
  config: CreateableTypeConfig<Source, Context>,
) => {
  // console.log(typeName, config.fields)
  return isEmpty(config.fields)
    ? ''
    : `
  ${createable} ${typeName} ${
        config.interface ? `implements ${config.interface} ` : ''
      }{
    ${reduceRecord(config.fields)}
  }
`
}

const generateNonEmpty = <Source, Context>(
  typeName: Resolvables,
  entries: CreateableTypeConfig<Source, Context>,
) =>
  isEmpty(entries.fields)
    ? ''
    : generateCreateableType('type', typeName, entries)

const generateEnums = (enums: Record<string, EnumTypesRecordConfig>) =>
  map(enums, (config, enumName) => {
    return config.values.length
      ? `enum ${enumName} { ${config.values
          .map(value => `${value}`)
          .join(' ')} }`
      : ''
  }).join('\n')

const generateScalars = (scalar: string[]) =>
  scalar.map(name => `scalar ${name}`).join('\n')

const generateFields = <Source, Context>(
  createable: string,
  createables: Record<string, CreateableTypeConfig<Source, Context>>,
) =>
  reduce(
    createables,
    (result, config, typeName) => {
      result.push(generateCreateableType(createable, typeName, config))
      return result
    },
    [],
  ).join('\n')

export const generateTypeDefs = <Source, Context>(
  resolvables: ResolvablesRecord<Source, Context>,
  types: CreateableTypesRecord<Source, Context>,
): ITypeDefinitions => {
  if (isEmpty(resolvables.Query)) throw new Error('Query cannot be empty')

  return `
    ${generateFields('interface', types.interface)}
    ${generateFields('input', types.input)}
    ${generateFields('type', types.type)}
    ${generateEnums(types.enum)}
    ${generateCreateableType('type', 'Query', resolvables.Query)}
    ${generateNonEmpty('Mutation', resolvables.Mutation)}
    ${generateNonEmpty('Subscription', resolvables.Subscription)}
    ${generateScalars(types.scalar)}
  `
}
