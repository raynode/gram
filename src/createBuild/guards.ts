import { BuildModeGenerator } from './types'

export const isBuildModeGenerator = <BuildMode, Result>(
  val: any,
): val is BuildModeGenerator<BuildMode, Result> => typeof val === 'function'
