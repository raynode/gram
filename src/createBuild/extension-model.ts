import { GraphQLNonNull, GraphQLType, isType } from 'graphql'
import { IFieldResolver } from 'graphql-tools'
import { reduce } from 'lodash'

import {} from '../attributeBuilder'
import { AttributeBuilder, ModelBuilder } from '../types'

import { isBuildModeGenerator } from './guards'
import { Build } from './types'

export interface WithAddModel<BuildMode> extends Build<BuildMode> {
  addModel: <Type, GQLType = Type>(
    model: ModelBuilder<BuildMode, Type, GQLType>,
  ) => void
}

const fakeWrapped = <BuildMode>(buildMode: BuildMode) =>
  ({
    id: 'string',
    getBaseModel: (name: string) => name,
    getModel: (name: string) => name,
    addModel: () => null,
    filterStrategy: null,
    buildMode,
    getScalar: (key: string) => key,
    pubSub: null,
  } as any)

const buildAttribute = <BuildMode>(
  buildMode: BuildMode,
  attr: AttributeBuilder<BuildMode, any, any>,
) => {
  const type = attr.field(fakeWrapped(buildMode))
  const gqlType = isType(type) ? type.toString() : type
  if (attr.listType) return `[${gqlType}!]!`
  if (!attr.nullable) return `${gqlType}!`
  return gqlType
}

export const addModel = <BuildMode>(baseBuild: Build<BuildMode>) => {
  const build = baseBuild as WithAddModel<BuildMode>
  build.addModel = modelBuilder => {
    const resolver: IFieldResolver<any, any> = modelBuilder.getResolver()

    const type = modelBuilder.isInterface() ? 'interface' : 'type'
    const typeConfig = {
      fields: reduce(
        modelBuilder.getAttributes(),
        (fields, attr, name) => {
          fields[name] = buildAttribute(build.buildMode, attr)
          return fields
        },
        {},
      ),
    }
    if (type === 'interface')
      build.addType(modelBuilder.name, 'interface', typeConfig)
    else
      build.addType(modelBuilder.name, 'type', {
        ...typeConfig,
        interface: modelBuilder.getInterfaces()[0],
      })
  }
  build.addQuery('test', 'String')
  return build
}
