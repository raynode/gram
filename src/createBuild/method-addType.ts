import {
  BuildModeArgsGenerator,
  BuildModeGenerator,
  CreateableTypes,
  Fields,
} from './types'

import { isBuildModeGenerator } from './guards'

export type Definable<BuildMode, Data> =
  | Data
  | BuildModeGenerator<BuildMode, Data>
export type DefinableFields<BuildMode> = Definable<BuildMode, Fields>

export type AddScalarArgsType = [string, 'scalar', never]
export type AddInterfaceArgsType<BuildMode> = [
  string,
  'interface',
  { fields: DefinableFields<BuildMode> }
]
export type AddInputArgsType<BuildMode> = [
  string,
  'input',
  { fields: DefinableFields<BuildMode> }
]
export type AddEnumArgsType<BuildMode> = [
  string,
  'enum',
  { values: Definable<BuildMode, string[]> }
]
export type AddObjectArgsType<BuildMode> = [
  string,
  'type',
  {
    fields: DefinableFields<BuildMode>;
    interface?: Definable<BuildMode, string>;
  }
]

export type AddArgs<BuildMode> =
  | AddScalarArgsType
  | AddInterfaceArgsType<BuildMode>
  | AddInputArgsType<BuildMode>
  | AddEnumArgsType<BuildMode>
  | AddObjectArgsType<BuildMode>

export type AddScalarType = (
  typeName: string,
  type: 'scalar',
  config?: {},
) => void
export type AddInterfaceType<BuildMode> = (
  typeName: string,
  type: 'interface',
  config: {
    fields: DefinableFields<BuildMode>;
  },
) => void
export type AddInputType<BuildMode> = (
  typeName: string,
  type: 'input',
  config: {
    fields: DefinableFields<BuildMode>;
  },
) => void
export type AddEnumType<BuildMode> = (
  typeName: string,
  type: 'enum',
  config: {
    values: Definable<BuildMode, string[]>;
  },
) => void
export type AddObjectType<BuildMode> = (
  typeName: string,
  type: 'type',
  config: {
    fields: DefinableFields<BuildMode>;
    interface?: Definable<BuildMode, string>;
  },
) => void

export type AddType<BuildMode> = AddScalarType &
  AddInterfaceType<BuildMode> &
  AddInputType<BuildMode> &
  AddEnumType<BuildMode> &
  AddObjectType<BuildMode>

export type AddArgsGenerator<BuildMode> = BuildModeArgsGenerator<
  BuildMode,
  AddArgs<BuildMode>
>

export const createAddType = <BuildMode>(
  buildMode: BuildMode,
  addScalarType: AddScalarType,
  addInterfaceType: AddInterfaceType<BuildMode>,
  addInputType: AddInputType<BuildMode>,
  addEnumType: AddEnumType<BuildMode>,
  addObjectType: AddObjectType<BuildMode>,
): AddType<BuildMode> & AddArgsGenerator<BuildMode> => (
  buildModeBenerator,
  ...args
) => {
  const typeInformation = isBuildModeGenerator<BuildMode, AddArgs<BuildMode>>(
    buildModeBenerator,
  )
    ? buildModeBenerator(buildMode)
    : ([buildModeBenerator, ...args] as AddArgs<BuildMode>)
  switch (typeInformation[1]) {
    case 'scalar':
      return addScalarType(...typeInformation)
    case 'interface':
      return addInterfaceType(...typeInformation)
    case 'input':
      return addInputType(...typeInformation)
    case 'enum':
      return addEnumType(...typeInformation)
    case 'type':
      return addObjectType(...typeInformation)
  }
}
