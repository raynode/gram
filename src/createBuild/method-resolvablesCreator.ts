import {
  AddResolvable,
  AddResolvableConfig,
  AddResolver,
  BuildModeGenerator,
  FieldType,
  GQLRecord,
  Resolvables,
} from './types'

import { IFieldResolver } from 'graphql-tools'

import { typeToString } from './utils'

import { isBuildModeGenerator } from './guards'

export const createBuildModeResolver = <BuildMode>(buildMode: BuildMode) => <
  Result
>(
  data: Result | BuildModeGenerator<BuildMode, Result>,
) => (isBuildModeGenerator<BuildMode, Result>(data) ? data(buildMode) : data)

export const resolvablesCreator = <BuildMode, Context>(
  buildMode: BuildMode,
  resolvables: Record<Resolvables, GQLRecord>,
  addResolver: AddResolver<Context>,
) => {
  const resolveBuildModeGenerator = createBuildModeResolver(buildMode)
  return <Source, Context, Type = FieldType>(
    typeName: Resolvables,
  ): AddResolvable => (name: string, type: any, configOrResolver = {}) => {
    const { args = {}, resolver = null } =
      typeof configOrResolver === 'function'
        ? { args: {}, resolver: configOrResolver }
        : {
            args: resolveBuildModeGenerator(configOrResolver.args),
            resolver: resolveBuildModeGenerator(configOrResolver.resolver),
          }
    const config = { args, resolver }
    if (isBuildModeGenerator<BuildMode, FieldType>(type)) {
      resolvables[typeName][name] = typeToString(
        resolveBuildModeGenerator(type),
      )
      addResolver(typeName, name, config)
    } else {
      resolvables[typeName][name] = typeToString(type)
      addResolver(typeName, name, config)
    }
  }
}
