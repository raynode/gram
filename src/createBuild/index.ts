import { IFieldResolver, makeExecutableSchema } from 'graphql-tools'

import { generateTypeDefs } from './generateTypeDefs'
import { createAddType } from './method-addType'
import { resolvablesCreator } from './method-resolvablesCreator'
import {
  AddResolver,
  Build,
  CreateableTypesRecord,
  GQLRecord,
  Resolvables,
  Resolvers,
} from './types'
import { createBuildModeResolver, fieldsToGQLRecord } from './utils'

import { GQLBUILDER } from '../types/constants'

export * from './extension-model'

export const createBuild = <BuildMode = null, Context = any>(
  buildMode?: BuildMode,
) => {
  const resolvables: Record<Resolvables, GQLRecord> = {
    Mutation: {},
    Query: {},
    Subscription: {},
  }
  const types: CreateableTypesRecord = {
    enum: {},
    input: {},
    interface: {},
    scalar: [],
    type: {},
  }
  const resolvers: Resolvers = {}
  const addResolver: AddResolver<Context> = (
    base,
    name,
    { args, resolver },
  ) => {
    resolvers[base] = resolvers[base] || {}
    resolvers[base][name] = resolver
  }

  const createResolvable = resolvablesCreator<BuildMode, Context>(
    buildMode,
    resolvables,
    addResolver,
  )

  const buildModeResolver = createBuildModeResolver(buildMode)
  const addInterfaceOrInputType = (typeName, createAble, config) =>
    (types[createAble][typeName] = {
      fields: fieldsToGQLRecord(buildModeResolver(config.fields)),
    })

  const addType = createAddType(
    buildMode,
    typeName => types.scalar.push(typeName),
    addInterfaceOrInputType,
    addInterfaceOrInputType,
    (typeName, createAble, config) =>
      (types.enum[typeName] = { values: buildModeResolver(config.values) }),
    (typeName, createAble, config) =>
      (types.type[typeName] = {
        fields: fieldsToGQLRecord(buildModeResolver(config.fields)),
        interface: buildModeResolver(config.interface),
      }),
  )

  const builder: Build<BuildMode, Context> = {
    type: GQLBUILDER,
    buildMode,
    addQuery: createResolvable('Query'),
    addMutation: createResolvable('Mutation'),
    addSubscription: createResolvable('Subscription'),
    addType,
    toSchema: () => makeExecutableSchema(builder.toTypeDefs()),
    toTypeDefs: () => ({
      typeDefs: generateTypeDefs(resolvables, types),
      resolvers,
    }),
  }
  return builder
}
