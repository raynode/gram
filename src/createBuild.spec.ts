import { graphql, GraphQLString } from 'graphql'
import { makeExecutableSchema } from 'graphql-tools'
import { createBuild } from './createBuild'

describe('createBuild', () => {
  it('should create a base object with methods to add everything for gql', () => {
    const build = createBuild()
    expect(build).toMatchSnapshot()
  })

  it('should be able to create a simple graphql interface', async () => {
    const build = createBuild()
    build.addQuery('test', GraphQLString, () => 'TEST')
    const { typeDefs, resolvers } = build.toTypeDefs()
    const schema = makeExecutableSchema({
      typeDefs,
      resolvers,
    })
    const result = await graphql({ schema, source: '{ test }' })
    expect(result).toMatchSnapshot()
  })

  it('should accept a buildMode function', async () => {
    const build = createBuild()
    build.addQuery('test', buildMode => GraphQLString, () => () => 'TEST')
    const { typeDefs, resolvers } = build.toTypeDefs()
    const schema = makeExecutableSchema({
      typeDefs,
      resolvers,
    })
    const result = await graphql({ schema, source: '{ test }' })
    expect(result).toMatchSnapshot()
  })
})
