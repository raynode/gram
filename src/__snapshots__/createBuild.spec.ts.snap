// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBuild should accept a buildMode function 1`] = `
Object {
  "data": Object {
    "test": "TEST",
  },
}
`;

exports[`createBuild should be able to create a simple graphql interface 1`] = `
Object {
  "data": Object {
    "test": "TEST",
  },
}
`;

exports[`createBuild should create a base object with methods to add everything for gql 1`] = `
Object {
  "addMutation": [Function],
  "addQuery": [Function],
  "addSubscription": [Function],
  "addType": [Function],
  "buildMode": undefined,
  "builder": "Build",
  "toSchema": [Function],
  "toTypeDefs": [Function],
}
`;
